File: extensions/journal/chronicler.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// externals
#include "external.h"
// namespace setup
#include "forward.h"


// add bindings for the chronicler, the object that holds the journal global state
void
pyre::journal::py::chronicler(py::module & m)
{
    // the chronicler interface
    py::class_<chronicler_t>(m, "Chronicler")
        // accessors
        // decor
        .def_property_static(
            "decor",
            // the getter
            [](py::object) -> chronicler_t::detail_type {
                // ask {chronicler_t}
                return chronicler_t::decor();
            },
            // the setter
            [](py::object, chronicler_t::detail_type decor) -> chronicler_t::detail_type {
                // set the new value and return the old
                return chronicler_t::decor(decor);
            },
            // the docstring
            "access the message decoration level")

        // detail
        .def_property_static(
            "detail",
            // the getter
            [](py::object) -> chronicler_t::detail_type {
                // ask {chronicler_t}
                return chronicler_t::detail();
            },
            // the setter
            [](py::object, chronicler_t::detail_type detail) -> chronicler_t::detail_type {
                // set the new value and return the old
                return chronicler_t::detail(detail);
            },
            // the docstring
            "access the maximum detail level")

        // device
        .def_property_static(
            "device",
            // the getter
            [](py::object) -> chronicler_t::device_type {
                // ask {chronicler_t}
                return chronicler_t::device();
            },
            // the setter
            [](py::object, chronicler_t::device_type device) -> void {
                // set the new device
                chronicler_t::device(device);
                // all done
                return;
            },
            // the docstring
            "access the default device")

        // global metadata
        .def_property_readonly_static(
            "notes",
            // the getter
            [](py::object) -> chronicler_t::notes_type & {
                // ask {chronicler_t}
                return chronicler_t::notes();
            },
            // the docstring
            "access the global metadata")

        // send output to a trash can
        .def_static(
            "quiet",
            // the implementation
            &chronicler_t::quiet,
            // the docstring
            "suppress all output from all channels")

        // all done
        ;

    // all done
    return;
}


// end of file

File: packages/journal/Chronicler.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


# for my metaclass
import pyre


# the singleton that owns the global state
class Chronicler(metaclass=pyre.patterns.singleton):
    """
    The manager of the journal global state
    """


    # public data
    notes = None
    device = None
    decor = 1
    detail = 1


    # metamethods
    def __init__(self, decor=decor, detail=detail, device=device, notes=notes, **kwds):
        # chain up
        super().__init__(**kwds)

        # the default decor
        self.decor = decor
        # the default detail
        self.detail = detail

        # the global metadata; can't be empty
        self.notes = notes if notes is not None else {
            "application": "journal",  # this key is required; applications should override
            }

        # if whoever initialized the journal did not express an opinion regarding the device
        if device is None:
            # grab the console
            from .Console import Console as cout
            # and instantiate it
            device = cout()
        # attach it
        self.device = device

        # all done
        return


# end of file
